### PR TITLE
Bug in type inference.

### LINKED ISSUE #1022: A compiler error on `Unification.hs` when accessing the field
I got a compiler error if I uncomment `var node = reversed_list.node` in the function `nonnull_list_nth()` to access the field. I don't quite understand the error message: `ddlog: Unification.teToType: non-constant type expression 'a50 CallStack (from HasCallStack): error, called at src/Language/DifferentialDatalog/Unification.hs:415:35 in differential-datalog-0.42.0-1PgGWEVSzEkJuq6cM15JM0:Language.DifferentialDatalog.Unification`. Hope I find another bug to improve ddlog or let me know the right way to write the function in my linked list implementation. Thanks!

```
function into_nonnull_vec(list: mut NonNullList<'N>): Vec<'N> {
	// Have to reverse the list first because `Vec<T>` does not support insertion
	// at the beginning of the vector
	var reversed_list = nonnull_list_reverse(list);
	// var node = reversed_list.node;
        // FIXME: A ugly workarond to please the compiler
	var node = nonnull_list_nth(reversed_list, 0).unwrap_or_default();
	var nxt = nonnull_list_nxt(reversed_list);
	// match (reversed_list.nxt) {
	match (nxt) {
		NonNullNxt { nxt_nonnull_list_ref } -> {
			var nxt_nonnull_list = deref(nxt_nonnull_list_ref);
			var vector = into_nonnull_vec(nxt_nonnull_list);
			vec_push(vector, node);
			vector
		},
		NULL -> {
			var vector = vec_empty();
			vec_push(vector, node);
			vector
		}
	}
}
```

Below is a quick implementation of linked list in DDLog that probably has the worst performance on list operations because of tons of `deref()` and ugly workarounds but I'm using this linked list implementation somewhere that does not require high performance or even incrementality. I appreciate any suggestion to enhance it even though you may not want it to be included in the standard library.

```
// The `NonNullList` must have at least one item in the list and can't be null
// `NonNullList` in Formula representation:
// LinkedList ::= new (node: N, nxt: LinkedList + NULL)
typedef NonNullList<'N> = NonNullList { node: 'N, nxt: NonNullNxt<'N> }
typedef NonNullNxt<'N> = NonNullNxt { nxt: Ref<NonNullList<'N>> } | NULL

function from_singleton_to_nonnull_list(item: 'N): NonNullList<'N> {
	NonNullList { item, NULL }
}

function from_nonnull_vec(vector: mut Vec<'N>): Option<NonNullList<'N>> {
	if (vec_len(vector) == 0) {
		None
	} else {
		var last = vector.pop().unwrap_or_default();
		var list_without_last_opt = from_nonnull_vec(vector);
		match (list_without_last_opt) {
			Some {list_without_last} -> {
				var nonnull_list = nonnull_list_append(list_without_last, last);
				Some { nonnull_list }
			},
			None -> { 
				Some { NonNullList { last, NULL } }
			}
		}
	}
}

function into_nonnull_vec(list: mut NonNullList<'N>): Vec<'N> {
	// Have to reverse the list first because `Vec<T>` does not support insertion
	// at the beginning of the vector
	var reversed_list = nonnull_list_reverse(list);
	// FIXME: A workarond to please the compiler
	// var node = reversed_list.node;
	var node = nonnull_list_nth(reversed_list, 0).unwrap_or_default();
	var nxt = nonnull_list_nxt(reversed_list);
	// match (reversed_list.nxt) {
	match (nxt) {
		NonNullNxt { nxt_nonnull_list_ref } -> {
			var nxt_nonnull_list = deref(nxt_nonnull_list_ref);
			var vector = into_nonnull_vec(nxt_nonnull_list);
			vec_push(vector, node);
			vector
		},
		NULL -> {
			var vector = vec_empty();
			vec_push(vector, node);
			vector
		}
	}
}

// A helper function for the temporary workaroud
function nonnull_list_nxt(list: NonNullList<'N>): NonNullNxt<'N> {
	list.nxt
}

function nonnull_list_nth(list: NonNullList<'N>, i: usize): Option<'N> {
	if (i == 0) {
		Some{ list.node }
	} else {
		match (list.nxt) {
			NonNullNxt { nxt_nonnull_list_ref } -> {
				var nxt_nonnull_list = deref(nxt_nonnull_list_ref);
				nonnull_list_nth(nxt_nonnull_list, i-1)
			},
			NULL -> { None }
		}
	}
}

function nonnull_list_append(list: mut NonNullList<'N>, item: 'N): NonNullList<'N> {
	match (list.nxt) {
		NonNullNxt { nxt_nonnull_list_ref } -> {
			var nxt_nonnull_list = deref(nxt_nonnull_list_ref);
			var nxt_nonnull_list_with_item = nonnull_list_append(nxt_nonnull_list, item);
			nxt_nonnull_list_with_item
		},
		NULL -> { 
			// Create a NonNullList with two items
			var nxt = NonNullNxt { ref_new(NonNullList { item, NULL }) };
			NonNullList { list.node, nxt }
		}
	}
}

function nonnull_list_reverse(list: mut NonNullList<'N>): NonNullList<'N> {
	match (list.nxt) {
		NonNullNxt { nxt_nonnull_list_ref } -> {
			var nxt_nonnull_list = deref(nxt_nonnull_list_ref);
			var reversed_nxt_nonnull_list = nonnull_list_reverse(nxt_nonnull_list);
			nonnull_list_append(reversed_nxt_nonnull_list, list.node)
		},
		NULL -> { list }
	}
}

function nonnull_list_map(list: NonNullList<'N>, f: function('N): 'M): NonNullList<'M> {
	var new_node = f(list.node);
	var mapped_nonnull_nxt = match (list.nxt) {
		NonNullNxt {nxt_nonnull_list_ref} -> {
			var nxt_nonnull_list = deref(nxt_nonnull_list_ref);
			var mapped_nonnull_list = nonnull_list_map(nxt_nonnull_list, f);
			NonNullNxt { ref_new(mapped_nonnull_list) }
		},
		NULL -> NULL
	};
	NonNullList { 
		new_node, 
		mapped_nonnull_nxt
	}
}

// `List` in Formula representation:
// NonNullLinkedList ::= new (node: N, nxt: LinkedList)
// LinkedList ::= NonNullLinkedList + NULL 
typedef List<'N> = List { node: 'N, nxt: ListNxt<'N> } | EMPTY
typedef ListNxt<'N> = ListNxt { nxt: Ref<List<'N>> }

function from_singleton_to_list(item: 'N): List<'N> {
	List { item, ListNxt {ref_new(EMPTY)} }
}

function from_vec(vector: mut Vec<'N>): List<'N> {
	if (vec_len(vector) == 0) {
		EMPTY
	} else {
		var last = vector.pop().unwrap_or_default();
		var list_without_last = from_vec(vector);
		list_append(list_without_last, last)
	}
}

function into_vec(list: mut List<'N>): Vec<'N> {
	var reversed_list = list_reverse(list);
	match (reversed_list) {
		List { node, ListNxt {nxt_list_ref} } -> {
			var nxt_list = deref(nxt_list_ref);
			var inner_vec = into_vec(nxt_list);
			vec_push(inner_vec, node);
			inner_vec
		},
		EMPTY -> { vec_empty() }
	}
}

function list_nth(list: List<'N>, i: usize): Option<'N> {
	match (list) {
		List { node, ListNxt {nxt_list_ref} } -> {
			if (i == 0) { Some {node} } 
			else {
				var nxt_list = deref(nxt_list_ref);
				list_nth(nxt_list, i-1)
			}
		},
		EMPTY -> { None }
	}
}

// Since `Ref<List<'N>>` is immutable we have to make a deep copy and append the new item
function list_append(list: mut List<'N>, item: 'N): List<'N> {
	match (list) {
		List { node, ListNxt { nxt_list_ref } } -> {
			var nxt_list = deref(nxt_list_ref);
			var nxt_list_with_item = list_append(nxt_list, item);
			List { node, ListNxt { ref_new(nxt_list_with_item) } }
		},
		EMPTY -> {
			List { item, ListNxt { ref_new(EMPTY) } }
		}
	}
}

function list_reverse(list: mut List<'N>): List<'N> {
	match (list) {
		List { node, ListNxt {nxt_list_ref} } -> {
			// TODO: Deref literally on every iteration is bad.
			var nxt_list = deref(nxt_list_ref);
			var reversed_nxt_list = list_reverse(nxt_list);
			list_append(reversed_nxt_list, node)
		},
		EMPTY -> EMPTY
	}
}

function list_map(list: List<'N>, f: function('N): 'M): List<'M> {
    match (list) {
		List { node, ListNxt {nxt_list_ref} } -> {
			var nxt_list = deref(nxt_list_ref);
			var mapped_nxt_list = list_map(nxt_list, f);
			List { f(node), ListNxt {ref_new(mapped_nxt_list)} }
		},
		EMPTY -> EMPTY
    }
}
```

### PR DESCRIPTION
**Pull request by ryzhyk**
Fixes a bug in lazy constraint expansion.  When performing type
inference for a field access expression (`x.f`), we generate a lazy
constraint with an branch per each struct type that has a field named `n`.
We expand the constraint as soon as we know the name of the struct, at
which point we can prune all incorrect options.  However, the code
incorrectly assumed that the field type is also fully concretized at
this point, which is generally not true, as type arguments may still be
unknown types.

The fix doesn't attempt to concretize the field completely, but instead
simply substitutes what we know about type arguments at this point.

Resolves #1022

### CODE REVIEW

File: test/datalog_tests/simple3.dl [hunk 1]
```diff
@@ -0,0 +1,208 @@
+/******************************************
```
Comment (mbudiu-vmw) [inline test/datalog_tests/simple3.dl:1]: This is just a test or is this supposed to be something we want to reuse?

File: test/datalog_tests/simple3.dl [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    NonNullList{ 1, NonNullNxt{ ref_new(NonNullList{ 2, NonNullNxt{ ref_new(NonNullList{ 3, NULL }) } }) } }
+}
+
+NonNullListTest("from_nonnull_vec([1,2,3,4,5]): ${from_nonnull_vec([1,2,3,4,5]).to_string_debug()}").
+NonNullListTest("test_non_null_list(): ${test_non_null_list().to_string_debug()}").
+NonNullListTest("into_nonnull_vec(test_non_null_list()): ${into_nonnull_vec(test_non_null_list()).to_string_debug()}").
+NonNullListTest("nonnull_list_append(test_non_null_list(), 4): ${nonnull_list_append(test_non_null_list(), 4).to_string_debug()}").
+NonNullListTest("nonnull_list_map(test_non_null_list(), +1): ${nonnull_list_map(test_non_null_list(), |x| x + 1).to_string_debug()}").
+
+// `List` in Formula representation:
```
Comment (mbudiu-vmw) [inline test/datalog_tests/simple3.dl:124]: is this stuff tested anywhere?
Or is it here just to trigger the generics bug?

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,25 @@ All notable changes to this project will be documented in this file.
...
+
+### New library functions
+
+- New `ddlog_std` library functions:
+
+    ```
+    /* Convert any DDlog type into a string in a programmer-facing,
+     * debugging context.  Implemented by calling the `Debug::fmt()`
+     * method of the underlying Rust type. */
+    extern function to_string_debug(x: 'T): string
```
Comment (mbudiu-vmw) [inline CHANGELOG.md:20]: should this be the default to_string method?

**Review by mbudiu-vmw — APPROVED** [review]

File: test/datalog_tests/simple3.dl (hunk 1 above)
Comment (ryzhyk) [reply test/datalog_tests/simple3.dl:1]: Just a test

File: test/datalog_tests/simple3.dl (hunk 2 above)
Comment (ryzhyk) [reply test/datalog_tests/simple3.dl:124]: Nope, the functionality is not tested. It's mostly to test type inference.  Ideally, there would be functional tests too, but I don't want to spend more time on this atm.

@qishen, if you ever finish this library and have some tests for it, please share them with us!

File: CHANGELOG.md (hunk 3 above)
Comment (ryzhyk) [reply CHANGELOG.md:20]: Problem with default `to_string()` implementation is that at the moment neither DDlog nor Rust support specialization, so we cannot easily have a blanket implementation that appliez to all types and then overload it for a particular type.

File: test/datalog_tests/simple3.dl (hunk 2 above)
Comment (qishen) [reply test/datalog_tests/simple3.dl:125]: Thanks for the comments and suggestions! I just found some issues with my current implementation of the linked list that returns the wrong result. I'll share it with you later once I fix it.

**Pull request merged**